### PR TITLE
03613 token freeze unfreeze precompile

### LINKED ISSUE #3613: HIP 514: Add token management functions for freeze flow
### Problem

The precompile flow is missing support for the freeze flow

### Solution

Add support for token freeze management flows

- isFrozen(address)
- tokenFreeze(address, address)
- tokenUnFreeze(address, address)

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by nikolovyanko**
**Description**:
This PR introduces token freeze, unfreeze write operations, and isFrozen read operation for precompiles.

 - New [FreezeLogic](https://github.com/hashgraph/hedera-services/blob/f31e38512de61df4d6e32e1da5e4a413a5c80854/hedera-node/src/main/java/com/hedera/services/txns/token/FreezeLogic.java) and [UnFreezeLogic](https://github.com/hashgraph/hedera-services/blob/f31e38512de61df4d6e32e1da5e4a413a5c80854/hedera-node/src/main/java/com/hedera/services/txns/token/UnFreezeLogic.java) making  the transition logic accessible by the HTSPrecompileContracts.
 - A new [AbstractFreezeUnfreezePrecompile](https://github.com/hashgraph/hedera-services/blob/f31e38512de61df4d6e32e1da5e4a413a5c80854/hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/impl/AbstractFreezeUnfreezePrecompile.java) class for freeze and unfreeze operations, extended by:
   - [FreezeTokenPrecompile](https://github.com/hashgraph/hedera-services/blob/f31e38512de61df4d6e32e1da5e4a413a5c80854/hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/impl/FreezeTokenPrecompile.java)
   - [UnfreezeTokenPrecompile](https://github.com/hashgraph/hedera-services/blob/f31e38512de61df4d6e32e1da5e4a413a5c80854/hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/impl/UnfreezeTokenPrecompile.java)

- A new [IsFrozenPrecompile](https://github.com/hashgraph/hedera-services/blob/f31e38512de61df4d6e32e1da5e4a413a5c80854/hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/impl/IsFrozenPrecompile.java) for IsFrozen read operation.

- New methods for encoding/decoding result/params for freeze calls introduced in [EncodingFacade](https://github.com/hashgraph/hedera-services/blob/4cd015f86e6c36c7bd5ec8567a81980ad046b001/hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/EncodingFacade.java) and [DecodingFacade](https://github.com/hashgraph/hedera-services/blob/4cd015f86e6c36c7bd5ec8567a81980ad046b001/hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/DecodingFacade.java)

**Related issue(s)**: 

Fixes #3613 

**Notes for reviewer**:

**Checklist**
- [x] Implementation
- [x] Unit tests 
- [x] E2E tests

### CODE REVIEW

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/TokenFreezeUnfreezeWrapper.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hedera.services.store.contracts.precompile.codec;
+
+import com.hederahashgraph.api.proto.java.AccountID;
+import com.hederahashgraph.api.proto.java.TokenID;
+
+public record TokenFreezeUnfreezeWrapper(TokenID token, AccountID account) {}
```
Comment (nikolovyanko) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/TokenFreezeUnfreezeWrapper.java:21]: Functions freeze, unfreeze, isFrozen receive the same tokenID and accountID arguments. Should there be a separate wrapper record for every function or is this approach with the single "TokenFreezeUnfreezeWrapper"  ok?

Comment (IvanKavaldzhiev) [reply hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/TokenFreezeUnfreezeWrapper.java:21]: We can have a common wrapper but separate building method for each precompile. This would make the code easier to maintain in the future. The methods can be:

- forFreeze
- forUnfreeze
- forIsFrozen

and they might reuse a common private method that initializes the wrapper.

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/AbiConstants.java [hunk 2]
```diff
@@ -93,6 +93,12 @@ private AbiConstants() {
     public static final int ABI_ID_ERC_OWNER_OF_NFT = 0x6352211e;
     // tokenURI(uint256 tokenId)
     public static final int ABI_ID_ERC_TOKEN_URI_NFT = 0xc87b56dd;
+    // isFrozen(address, address)
+    public static final int ABI_IS_TOKEN_FROZEN = 0x46de0fb1;
+    // freezeToken(address, address)
+    public static final int ABI_FREEZE_TOKEN = 0x5b8f8584;
+    // unfreezeToken(address, address)
+    public static final int ABI_UNFREEZE_TOKEN = 0x52f91387;
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/AbiConstants.java:101]: In order to be compliant with the other signature comments you can add variable names, not only types.

Basically the signature for `isFrozen` is `isFrozen(address,address)` but to have more context about the function and parameters we also have added parameter names like `from`, `to`, `tokenId`, `value` to have better understanding about the function. These parameter names, however, are not part of the signature itself and are not used to calculate the function selector.

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/DecodingFacade.java [hunk 3]
```diff
@@ -844,6 +866,42 @@ public TokenCreateWrapper decodeNonFungibleCreateWithFees(
...
+                decodeFunctionCall(
+                        input, FREEZE_TOKEN_FUNCTION_SELECTOR, FREEZE_TOKEN_ACCOUNT_DECODER);
+
+        final var tokenID = convertAddressBytesToTokenID(decodedArguments.get(0));
+        final var accountID =
+                convertLeftPaddedAddressToAccountId(decodedArguments.get(1), aliasResolver);
+        return TokenFreezeUnfreezeWrapper.forFreeze(tokenID, accountID);
+    }
+
+    public TokenFreezeUnfreezeWrapper decodeUnFreeze(
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/DecodingFacade.java:893]: decodeUnFreeze -> decodeUnfreeze

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/impl/AbstractFreezeUnfreezePrecompile.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+import com.hedera.services.store.contracts.precompile.utils.KeyActivationUtils;
+import com.hedera.services.store.contracts.precompile.utils.PrecompilePricingUtils;
+import com.hedera.services.store.models.Id;
+import java.util.Objects;
+import org.hyperledger.besu.evm.frame.MessageFrame;
+
+/* --- Constructor functional interfaces for mocking --- */
+public abstract class AbstractFreezeUnfreezePrecompile extends AbstractWritePrecompile {
+
+    private final boolean isFreeze;
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/impl/AbstractFreezeUnfreezePrecompile.java:39]: isFreeze -> isFrozen

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/impl/AbstractFreezeUnfreezePrecompile.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        final var accountStore = infrastructureFactory.newAccountStore(ledgers.accounts());
+        final var tokenStore =
+                infrastructureFactory.newTokenStore(
+                        accountStore,
+                        sideEffects,
+                        ledgers.tokens(),
+                        ledgers.nfts(),
+                        ledgers.tokenRels());
+
+        final var freezingLogic =
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/impl/AbstractFreezeUnfreezePrecompile.java:92]: freezingLogic -> freezeLogic

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/InfrastructureFactory.java [hunk 6]
```diff
@@ -200,6 +202,16 @@ public DeleteAllowanceLogic newDeleteAllowanceLogic(
         return new DeleteAllowanceLogic(accountStore, tokenStore);
     }
 
+    public FreezeLogic newFreezeLogic(
+            final AccountStore accountStore, final TypedTokenStore tokenStore) {
+        return new FreezeLogic(tokenStore, accountStore);
+    }
+
+    public UnFreezeLogic newUnFreezeLogic(
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/InfrastructureFactory.java:210]: UnFreezeLogic -> UnfreezeLogic

File: hedera-node/src/main/java/com/hedera/services/txns/token/TokenFreezeUnfreezeLogic.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+import com.hedera.services.store.models.Id;
+import com.hederahashgraph.api.proto.java.ResponseCodeEnum;
+import com.hederahashgraph.api.proto.java.TransactionBody;
+
+public interface TokenFreezeUnfreezeLogic {
+    ResponseCodeEnum validate(TransactionBody txnBody);
+
+    void doFreezeUnfreeze(Id targetTokenId, Id targetAccountId);
+}
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/txns/token/TokenFreezeUnfreezeLogic.java:26]: The design would be cleaner if you remove the interface and keep 2 separate `Logic` implementations -> `FreezeLogic`, `UnfreezeLogic` without having them extend a common class/interface.

Having `FreezeLogic` with `doFreezeUnfreeze` method may imply that there is an option to do `unfreeze` operation inside a `FreezeLogic` and vice versa. This is confusing.

**Review by IvanKavaldzhiev — COMMENTED** [review]
Update HederaTokenService.sol and IHederaTokenService.sol files with the new methods.

**Review by IvanKavaldzhiev — COMMENTED** [review]
Add test-clients/src/main/resource/contract/contracts/FreezeUnfreezeContract/FreezeUnfreezeContract.json file

File: hedera-node/src/test/java/com/hedera/services/contracts/sources/TxnAwareEvmSigsVerifierTest.java [hunk 8]
```diff
@@ -174,6 +175,35 @@ void supplyKeyFailsWhenTokensLedgerIsNull() {
                                 true, PRETEND_TOKEN_ADDR, PRETEND_SENDER_ADDR, ledgers));
     }
 
+    @Test
+    void throwsIfAskedToVerifyTokenWithoutWipeKey() {
```
Comment (mustafauzunn) [inline hedera-node/src/test/java/com/hedera/services/contracts/sources/TxnAwareEvmSigsVerifierTest.java:179]: Would you mind changing name to throwsIfAskedToVerifyTokenWithoutFreezeKey

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/FreezeUnfreezeTokenPrecompileSuite.java [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+
+    @Override
+    protected Logger getResultsLogger() {
+        return log;
+    }
+
+    @Override
+    public List<HapiApiSpec> getSpecsInSuite() {
+        return allOf(
+                List.of(
```
Comment (mustafauzunn) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/FreezeUnfreezeTokenPrecompileSuite.java:94]: Wouldn't it better to shorten this to:
return List.of(...)

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/FreezeUnfreezeTokenPrecompileSuite.java [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+                                                                asAddress(accountID.get()))
+                                                        .logged()
+                                                        .payingWith(ACCOUNT)
+                                                        .gas(GAS_TO_OFFER),
+                                                contractCallLocal(
+                                                        FREEZE_CONTRACT,
+                                                        IS_FROZEN_FUNC,
+                                                        asAddress(vanillaTokenID.get()),
+                                                        asAddress(accountID.get())))))
+                .then();
```
Comment (mustafauzunn) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/precompile/FreezeUnfreezeTokenPrecompileSuite.java:344]: I think we should add here childRecordsCheck

**Review by mustafauzunn — COMMENTED** [review]
LGTM

**Review by stoqnkpL — DISMISSED** [review]
LGTM!

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/InfrastructureFactory.java [hunk 11]
```diff
@@ -197,7 +199,7 @@ public ViewExecutor newViewExecutor(
             final MessageFrame frame,
             final ViewGasCalculator gasCalculator,
             final StateView stateView,
-            final WorldLedgers ledgers) {
+            WorldLedgers ledgers) {
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/InfrastructureFactory.java:202]: Why is `final` removed?

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/EncodingFacade.java [hunk 12]
```diff
@@ -79,6 +79,7 @@ public class EncodingFacade {
             TupleType.parse(INT_BOOL_PAIR_RETURN_TYPE);
     private static final TupleType getTokenDefaultKycStatusType =
             TupleType.parse(INT_BOOL_PAIR_RETURN_TYPE);
+    private static final TupleType isTokenFrozenType = TupleType.parse(INT_BOOL_PAIR_RETURN_TYPE);
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/codec/EncodingFacade.java:82]: This type is already defined in `ParsingConstants`

File: hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/proxy/ViewExecutor.java [hunk 13]
```diff
@@ -53,7 +54,7 @@ public ViewExecutor(
             final DecodingFacade decoder,
             final ViewGasCalculator gasCalculator,
             final StateView stateView,
-            final WorldLedgers ledgers) {
+            WorldLedgers ledgers) {
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/proxy/ViewExecutor.java:57]: Why is `final` removed?

**Review by IvanKavaldzhiev — APPROVED** [review]
LGTM

**Review by stoqnkpL — APPROVED** [review]

**Review by Nana-EC — APPROVED** [review]
LG
